Task02

kelbor-khal opened the pull request:
Ночные размышления...

kelbor-khal committed:
Conspect lecture's

theory base

kelbor-khal committed:
added function strnlen()

code UAL function strnlen()

kelbor-khal committed:
added function strncpy(), comment strnlen

Написал построчные комментарии на русском, чтобы объяснить, как я представляю работу функции

kelbor-khal committed:
added function strstr()

Не уверен, что это работает. Даже теоретически!

Toliak commented on Filin_IA/task_01/strnlen.s:
```diff
@@ -0,0 +1,31 @@
+loop_strnlen:
+    mov al, byte [rbx]      ; выгрузка байта
+    add rcx, 1              ; инкрементирование счетчика
+    add rbx, rcx            ; переход к новому адресу
```
Тут скорее `add rbx, 1`, иначе на второй итерации будет `rbx = str + 1`, на третьей`rbx = (str + 1) + 2` и т.д.

Toliak commented on Filin_IA/task_01/strnlen.s:
```diff
@@ -0,0 +1,31 @@
+    jnz loop_strnlen 
+
+return_value:
+    sub rbx, rcx            ; определение количества байт
```
Оно не сработает корректно из-за того, что `rcx` используется в цикле как счетчик.
Скорее всего ожидалось `offset_str_ptr - original_str_ptr`, а получилось `offset_str_ptr - loop_index`

Toliak commented on Filin_IA/task_01/strnlen.s:
```diff
@@ -0,0 +1,31 @@
+    add rcx, 1              ; инкрементирование счетчика
+    add rbx, rcx            ; переход к новому адресу
+
+    cmp rcx, r14            ; сравнение с заданным значением счетчика 
```
`strnlen("asdasdasd", 0)` корректно отработает?
Рассмотрим:
```c
// loop_strnlen
al = *rbx;    // ok
rcx += 1;    // стало 1 (было 0)
rbx += rcx;  // не ok, описал выше почему. К текущему примеру не относится
// rcx == r14 ?  1 == 0? Нет, значит продолжаем цикл, хотя счетчик уже переполнился...
```

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,64 @@
+strstr:
+    push r9
+    push r10            
+    push r11
```
Этих ребят (r9, r10, r11), согласно [Microsoft x64 calling convention](https://en.wikipedia.org/wiki/X86_calling_conventions#Microsoft_x64_calling_convention), сохранять не надо

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,64 @@
+loop_lenght_search:
+    mov al, [r8]
+    add r9, 1
+    add r8, r9
```
Мне кажется тут тоже (как и в strnlen) нужно `add r8, 1` (или `inc r8`) делать

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,64 @@
+; r9 хранит в себе длину строки, которая должна входить в поле поиска
+
+loop_check:
+    mov r10, byte [rbx]              ; загружаем байт по указателю из поля поиска
+    mov r11, byte [r12]              ; загружаем байт по указателю искомой строки
```
64-битному регистру нельзя присвоить 8-битное значение.
Здесь и далее (на 30, 34, 38 строке) скорее всего имелось в виду `r10b`, `r11b` (т.е. 8-битная часть соответствующего регистра)

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,64 @@
+    cmp r10, 0
+    jz  return_value_not
+
+; r11 равняется нулю если вся строка прошла проверку 
```
Лайк за комменты 👍

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,64 @@
+    cmp r10, r12                     ; сравнение элементов строк
+    jnz loop_increment               ; если неравны
+                                     ; если элементы равны 
+    add rbx, rcx                     ; сдвигаем адрес в строке поиска
+    add r12, rcx                     ; сдвигаем адрес в искомой строке
```
Все еще кажется, что здесь нужен `inc rbx` и `inc r12`, т.к. иначе оно будет некорректно шагать

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,64 @@
+    jmp loop_check                   ; возврат к началу цикла
+
+return_value_ok:   
+    mov rax, [ rbx - r9]             ; текущее значение rbx минус длина искомой строки
```
К сожалению, [такой команды не существует](http://shell-storm.org/online/Online-Assembler-and-Disassembler/?inst=mov+rax%2C+%5Brbx+-+r9%5D&arch=x86-64&as_format=inline#assembly)

Плюс тут скорее всего имелось в виду `(char*)rax = (char*)rbx - (size_t)r9`, а это `lea`, а не `mov`

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,64 @@
+
+loop_increment
+    add rbx, rcx                     ; сдвигаем адрес в строке поиска
+    mov r12, rdx                     ; возвращаем к исходному значению
```
Возвращаться к исходному значению в случае несовпадения символов -- ок, согласен.
Но мы ведь сюда попадем даже если элементы равны (т.е. 42 строка, потом 45 и 46)

Toliak commented on Filin_IA/task_01/strncpy.s:
```diff
@@ -0,0 +1,39 @@
+    xor rcx, rcx            ; зануление счетчика
+
+loop_strncpy:
+    mov al, byte [r14]      ; загрузка одного байта из источника
```
`al` это часть `rax`, а в `rax` лежит возвращаемое значение (9 строка). Кажется, функция вернет что-то не то, что ожидалось :)

Toliak commented on Filin_IA/task_01/strncpy.s:
```diff
@@ -0,0 +1,39 @@
+loop_strncpy:
+    mov al, byte [r14]      ; загрузка одного байта из источника
+    add rcx, 1              ; инкрементирование счетчика
+    add r14, rcx            ; переход к новому адресу источника
```
База: `rcx = 0`, `r14 = base_str`.
1-я итерация: `rcx = 1`, `r14 = base_ptr + 1`
2-я итерация: `rcx = 2`, `r14 = (base_ptr + 1) + 2`
3-я итерация: `rcx = 3`, `r14 = ((base_ptr + 1) + 2) + 3`
и т.д.

Возможно, здесь имелось в виду `inc r14`

Toliak commented on Filin_IA/task_01/strncpy.s:
```diff
@@ -0,0 +1,39 @@
+    add rcx, 1              ; инкрементирование счетчика
+    add r14, rcx            ; переход к новому адресу источника
+
+    cmp rcx, r8             ; сравнение с заданным значением счетчика
```
Корректно ли отработает данный участок кода, если при вызове функции был передан `num = 0`?

Toliak commented on Filin_IA/task_01/strncpy.s:
```diff
@@ -0,0 +1,39 @@
+    test al, al             ; проверка на наличие нуль-символа
+    jz  filling_zeros       ; при равенстве 0 переход по метке
+                            ; если данные не равнются 0 и счетчик не переполнен
+    str al, [rbx]           ; загрузка данных по адресу строки назначения
```
Команды не существует :( Вероятно, подразумевалось `mov byte [rbx], al`

Toliak commented on Filin_IA/task_01/strncpy.s:
```diff
@@ -0,0 +1,39 @@
+
+filling_zeros:              ; заполнение нулями оставшийся объем строки
+
+    str 0, [rbx]            ; загрузка данных по адресу строки назначения
```
Команды не существует :( Вероятно, подразумевалось `mov byte [rbx], 0`

Toliak commented on Filin_IA/task_01/strncpy.s:
```diff
@@ -0,0 +1,39 @@
+    jz  filling_zeros       ; при равенстве 0 переход по метке
+                            ; если данные не равнются 0 и счетчик не переполнен
+    str al, [rbx]           ; загрузка данных по адресу строки назначения
+    add rbx, rcx            ; переход к новому адресу назначения
```
Аналогичный комментарий, как к 17 строке

Toliak replied:
Вопрос с тем, что тут нужен `inc rbx` остается открытым

Toliak reviewed (changes requested):
Некоторые комментарии пометил как "не критичные", их исправлять не обязательно.
Если я где-то ошибся или что-то не увидел -- можете писать ответом на соответствующие комментарии.

kelbor-khal committed:
fixed the function strnlen()

Исправил 1:
Теперь инкрементируется базовый регистр, отдельно от счетчика;

Исправил 2: 
Инкрементирование сетчика происходит после этапа сравнения его  с входным количеством символов;

Исправил 3:
Данная строчка вообще оказалась лишней :sub rbx, rcx.

kelbor-khal committed:
fixed the function strmcpy()

Исправил (без индекса):
Теперь указатель на скопированную строку сперва кладется в r13, а в самом конце укладывается в rax;

Исправил 1:
Аналогично предидущему коммиту, теперь инкрементируется адрес отдельно от счетчика;

Исправил 2:
Аналогично предидущему коммиту, теперь счетчик инкрементируется после этапа сравнения его значения с входными данными;

Исправил 3, 4:
Функция str заменена функцей mov byte; 

Исправил 5:
Аналогично исправлению 1.

kelbor-khal committed:
fixed the function strstr()

Исправил (без индекса):
Убрал сохранение в стек r9, r10, r11;

Исправил 1:
Исправил сдвиг адреса ( больное место, как и во всех предидущих программах). Добавил вычитание нуль-символа из длины искомой строки;

Исправил 2:
Исправил используемую размерность региистров r10 и r11 с 64 на 8;

Исправил 3:
Аналогично 1 исправлению;

Исправил 4:
Добавил функцию lea;

Исправил 5:
Поменял логику проверки на совпадения. Теперьв случае совпадения инкрементируется счетичк rcx, который вначале каждого прохода цикла сравнивается с длиной искомой строки и в случае совпадения выходит. Если символы не совпадают, то счетчик rcx обнуляется.

kelbor-khal committed:
added the function chksum()

Возможно запутался в размерности регистров

kelbor-khal committed:
Merge branch 'Filin_IA_task02' into Filin_Task02

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,71 @@
+loop_lenght_search:
+;------------ (исправил-1) ------------
+    mov al, [r8]
+    add r8, 1
```
Откуда тут появился `r8` и какова его цель?

UPD: посмотрел ниже, увидел комментарий
Получается, если я правильно понял, в начале функции был пропущен `mov r8, rdx`

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,71 @@
+
+strstr:
+
+    push r12
```
`push rbx` еще нужен (и pop в конце), т.к. он тоже используется и является non-volatile (по calling convention'у Microsoft x64)

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,71 @@
+; calling cinvention x86 OS Windows
+; argument's: rcx, rdx, r8, r9
+; return : rax
+; function strncpy ( const char * string1, const char * string2 )
```
(Не критично): я надеюсь, имелось в виду strstr, а не strncpy 😅

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,71 @@
+
+; если счетчик rcx равняется длине строки, то поиск завершен успешно
+    cmp r9, rcx
+    jz  return_value_ok
```
(Не критично): работать будет, но зачем так сложно?) Можно же проверить: если `r11b` содержит 0, то искомая строка закончилась

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,71 @@
+    cmp r9, rcx
+    jz  return_value_ok
+
+    cmp r10b, r12                    ; сравнение элементов строк
```
Возможно имелось в виду `cmp r10b, r11b`

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,71 @@
+    add rbx, 1                       ; сдвигаем адрес в строке поиска
+    add r12, 1                       ; сдвигаем адрес в искомой строке        
+    add rcx, 1
+;------------ (исправил-3) ------------
```
Тут точно не должно быть никакого перехода (jmp)?
Если пойти дальше, то оно еще раз прибавит 1 к rbx и начнет `loop_increment` выполнять...

Похоже, что надо прыгать на `loop_check`...

kelbor-khal replied:
Стыдно((

Toliak commented on Filin_IA/task_01/strstr.s:
```diff
@@ -0,0 +1,71 @@
+return_value_ok:
+
+;------------ (исправил-4) ------------
+    lea rax, [ rbx - r9]             ; текущее значение rbx минус длина искомой строки
```
Давайте все-таки сделаем существующей эту команду следующим образом:
```asm
mov rax, rbx
sub rax, r9
```

kelbor-khal replied:
а почему lea не будет работать?

Toliak replied:
[Эта команда не существует](http://shell-storm.org/online/Online-Assembler-and-Disassembler/?inst=lea+rax%2C+%5Brbx+-+r9%5D&arch=x86-64&as_format=inline#assembly)

Toliak commented on Filin_IA/task_01/strnlen.s:
```diff
@@ -0,0 +1,39 @@
+return_value:
+
+;------------ (исправил-3) ------------
+    add rcx, -1             ; вычитание нуль-символа
```
Кейс с `strnlen("asdasdasd", 0)` упадет на этом месте.
Давайте сделаем на 21-й строке `jz` на 34-ю строку

Toliak commented on Filin_IA/task_01/chksum.s:
```diff
@@ -0,0 +1,41 @@
+
+return_result:
+
+    mov eax, r9
```
`mov eax, r9d`

Toliak commented on Filin_IA/task_01/chksum.s:
```diff
@@ -0,0 +1,41 @@
+
+    mov al, byte [rbx]      ; загрузка одного байта из источника
+    add rbx, 1              ; двигаем адрес
+    add r9, al              ; прибавляем элемент массива
```
[Такой команды не существует](http://shell-storm.org/online/Online-Assembler-and-Disassembler/?inst=add+r9%2C+al&arch=x86-64&as_format=inline#assembly)

Здесь надо использоватьть `movzx eax, byte ptr [rbx]` (оно занулит весь (64 бита) регистр, а потом переложит в него 8 бит)
А потом `add r9d, eax` (r9d, т.к. нужно только 32 бита)

Toliak commented on Filin_IA/task_01/chksum.s:
```diff
@@ -0,0 +1,41 @@
+    add rbx, 1              ; двигаем адрес
+    add r9, al              ; прибавляем элемент массива
+
+    cmp rcx, rdx            ; сравнение счетчика с количеством элементов массива
```
По условию должен быть `edx`
И, соответственно, можно `ecx` вместо `rcx` использовать

Toliak commented on Filin_IA/task_01/strncpy.s:
```diff
@@ -0,0 +1,58 @@
+
+;------------ (исправил-5) ------------
+    add rbx, 1            ; переход к новому адресу назначения
+    jmp loop_strncpy        ; переход к началу функции       
```
`jmp filling_zeros` имелось в виду (плюс проверка счетчика)? Иначе оно ведь опять копировать пойдет...

Toliak reviewed (changes requested):
Стало лучше, но есть еще мелкие баги, иногда критичные...
Некоторые отметил как "не критично", их исправлять не надо

kelbor-khal committed:
fixed v.2 function strncpy()

Исправил:
- смещение адреса;
- добавил заполнение нулями, если в считываемой строке появился нуль-символ, а количество символов еще не достигнуто.

kelbor-khal committed:
fixed v.2 function strstr()

Исправил:
- выгрузил в стек rbx;
- теперь я сравниваю то что выгрузил, а не то что мне хочется)) r10b и r11b
- добавил прыжок к началу цикла;
- убрал lea и заменил на рекомендованную комбинацию.

Toliak commented:
Еще один момент
Переименуйте директорию `task01` в `task02`

kelbor-khal committed:
fixed function chksum()

Исправил:
- заменил mov на movzx;
- заменил rcx и rdx на ecx и edx ;
- r9d!.

kelbor-khal committed:
fixed v.2 function strnlen()

Исправил:
- теперь сразу выдаем результат.

kelbor-khal committed:
Rename dir

Переименовал директорию

kelbor-khal committed:
Merge branch 'Filin_Task02' into Filin_IA_task02

Toliak commented on Filin_IA/task_02/strstr.s:
```diff
@@ -0,0 +1,73 @@
+
+loop_increment
+    add rbx, 1                       ; сдвигаем адрес в строке поиска
+    mov r12, rdx                     ; возвращаем к исходному значению
```
Есть один неприятный момент
Тут бы еще сделать сброс `rbx`, но не полный сброс, а сброс до того момента, как начали обходить искомую строку.
[Пример случая, на котором сломается без этого сброса](https://github.com/xthebat/iu4-2k22-is/pull/10#discussion_r820101161)

kelbor-khal replied:
Каждый раз, когда у нас символы совпадают инкрементируются адреса искомой строки и строки, в которой ищем, а так же счетчик. В момент, когда проверив, какую-то часть строки, выгруженные символы не равны между собой, мы из текущего адреса rbx вычитаем  значение счетчика, потом инкрементируем rbx, а потом сбрасываем счетчик (rcx). Таким образом получается, что мы вернулись к тому моменту, где только начали проверять строку, а потом сдвинули адрес к следующему элементу.

Возможна следующая реализация:
53    sub rbx, rcx
54    add rbx, 1
55    mov r12, rdx
56    mov rcx, 0 
57    jmp loop_check

Вроде так)

Toliak replied:
Пойдет.
Еще, как вариант, сохранить "значение для восстановления rbx" в какой-нибудь регистр, а, в случае match-fail'а, достать сохраненное значение

Toliak commented on Filin_IA/task_02/chksum.s:
```diff
@@ -0,0 +1,41 @@
+    cmp ecx, edx               ; сравнение счетчика с количеством элементов массива
+    jz return_result
+
+    add rcx, 1                 ; инкрементирование счетчика
```
Упс, счетчик тоже `int` по условию CCC:

Toliak commented on Filin_IA/task_02/chksum.s:
```diff
@@ -0,0 +1,41 @@
+
+    mov rbx, rcx            ; запись в базовый регистр
+    xor rcx, rcx            ; обнуление rcx для последующей его работы как счетчика
+    xor r9, r9              ; а вдруг там, что-то лежало 
```
(Не критично): Не обязательно сбрасывать весь `r9`, если используется `r9d`. С другой стороны, хуже не станет :)

Toliak reviewed (changes requested):
- `chksum`: один мелкий фикс и будет закрыто
- `strstr`: один момент с логикой. Есть два варианта: пофиксить код; либо прям код не фиксить (все таки задание скорее "познакомиться с ассемблером"), но написать в комментарии возможный вариант решения.

Остальное норм

kelbor-khal committed:
fixed v.3 function chksum()

Теперь все что должно быть int стало int

kelbor-khal committed:
Merge branch 'Filin_IA_task02' into Filin_Task02

kelbor-khal committed:
Merge branch 'Filin_Task02' into Filin_IA_task02

Toliak reviewed (approved):
Баллы: 3/3 + 2/2
- `strncpy` +1
- `strnlen` +1
- `strstr` +1
- `chksum` +2

Нужен аппрув баллов от Алексея Алексеевича.
